### DISCUSSION TITLE
How often do I need to retrain IVF_FLAT  IVF_HNSW  ANNOY indexes for a lot of insertion?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18200] [Q&A and General discussion] [answered]
I want to play with different types of indexs, but can't understand how often do I need to train them?
My workload consist of a lot if inserts.
There are constant inserts.

Do these vectors even need a retraining?

**Comment by martinenkoEduard** [comment]
I just tried to create IVF_FLAT index (using Attu) with "nlist": 16384 to my collection with 5~ millions of entries.

```
print(fmt.format("Start Creating index IVF_FLAT"))
index = {
    "index_type": "IVF_FLAT",
    "metric_type": "L2",
    "params": {"nlist": 16384},
}
```

after that I tried to query one embedding

```
start_time = time.time()
search_param = {
        "data": [res[0]['embeddings']],
        "anns_field": "embeddings",
        "param": {"metric_type": "L2", "params": {"nprobe": 2}},
        "limit": 20}
results = collection.search(**search_param)

```

But query time was exactly the same like for just usual FLAT index.
changing nprobe: 2 dind't have any effect also.

**Comment by martinenkoEduard** [comment]
After release

i get following error if i try to load 
RPC error: [wait_for_loading_collection], <BaseException: (code=1, message=collection 434469916804382721 has not been loaded to memory or load failed)>, <Time:{'RPC start': '2022-07-10 03:49:40.880931', 'RPC error': '2022-07-10 03:55:19.906258'}>
RPC error: [load_collection], <BaseException: (code=1, message=collection 434469916804382721 has not been loaded to memory or load failed)>, <Time:{'RPC start': '2022-07-10 03:49:40.436106', 'RPC error': '2022-07-10 03:55:19.907319'}>
Traceback (most recent call last):
  File "/home/administrator/code/load.py", line 18, in <module>
    collection.load()
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/orm/collection.py", line 453, in load
    conn.load_collection(self._name, timeout=timeout, **kwargs)
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/decorators.py", line 56, in handler
    raise e
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/decorators.py", line 41, in handler
    return func(self, *args, **kwargs)
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/decorators.py", line 74, in handler
    raise e
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/decorators.py", line 70, in handler
    return func(*args, **kwargs)
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/client/grpc_handler.py", line 669, in load_collection
    self.wait_for_loading_collection(collection_name, timeout)
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/decorators.py", line 56, in handler
    raise e
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/decorators.py", line 41, in handler
    return func(self, *args, **kwargs)
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/decorators.py", line 74, in handler
    raise e
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/decorators.py", line 70, in handler
    return func(*args, **kwargs)
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/client/grpc_handler.py", line 689, in wait_for_loading_collection
    return self._wait_for_loading_collection_v2(collection_name, timeout)
  File "/home/administrator/anaconda3/lib/python3.9/site-packages/pymilvus/client/grpc_handler.py", line 713, in _wait_for_loading_collection_v2
    raise BaseException(response.status.error_code, response.status.reason)
pymilvus.client.exceptions.BaseException: <BaseException: (code=1, message=collection 434469916804382721 has not been loaded to memory or load failed)>

**Comment by xiaofan-luan** [comment]
> I want to play with different types of indexs, but can't understand how often do I need to train them? My workload consist of a lot if inserts. There are constant inserts.
> 
> Do these vectors even need a retraining?

You don't need to retrain.
The way you implement your code:
Create collection
Create index you want
load
insert entities/search

**Reply by martinenkoEduard** [reply]
What if I need to constantly add new data?
Will it work?

**Comment by yhmo** [comment, marked as answer]
You can do some experiments with this script:

```
import time
import numpy as np
from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection, Index,
    utility,
)

rng = np.random.default_rng(seed=11)
dim = 128
vector_field_name = "face"

def create_collection():
    fields = [
        FieldSchema(name="pk", dtype=DataType.INT64, is_primary=True, auto_id=False),
        FieldSchema(name="age", dtype=DataType.INT64),
        FieldSchema(name=vector_field_name, dtype=DataType.FLOAT_VECTOR, dim=dim)
    ]

    schema = CollectionSchema(fields)

    print("Create collection")
    collection = Collection("Test", schema, consistency_level="Eventually")
    return collection

def insert(collection):
    print("Start inserting entities")
    for i in range(500):
        print("insert 10000 entities", i)
        num_entities = 10000
        entities = [
            [i for i in range(num_entities)],
            [k for k in range(num_entities)],
            rng.random((num_entities, dim)),
        ]
        insert_result = collection.insert(entities)

    print(f"Number of entities in collection: {collection.num_entities}")

def load_collection(collection):
    print("Load collection")
    collection.load()

def release_collection(collection):
    print("Release collection")
    collection.release()

def search(collection):
    print("Searching")
    start = time.time()
    search_param = {
        "data": rng.random((1, dim)),
        "anns_field": vector_field_name,
        "param": {"metric_type": "L2", "params": {"nprobe": 2}},
        "limit": 10,
    }
    results = collection.search(**search_param)
    end = time.time()
    print("Search time cost:", (end-start)*1000, "ms")
    print(f"Search result:\n-{results}")

def create_index(collection):
    print("Start building index")
    index_params = {"index_type": "IVF_FLAT", "metric_type": "L2", "params": {"nlist": 512}}
    index = Index(collection, vector_field_name, index_params)

def check_index_progress(collection):
    state = utility.index_building_progress(collection_name=collection.name)
    print("Index building state:", state)

if __name__ == '__main__':
    print("Start connecting to Milvus")
    connections.connect("default", host="localhost", port="19530")

    #create a collection
    collection = create_collection()

    # #insert 5M entities
    insert(collection)

    #search without index, load() must be called before search/qiery
    load_collection(collection)
    search(collection)

    #build index
    create_index(collection)
    check_index_progress(collection)

    #search with index, release the original data, load index data
    release_collection(collection)
    load_collection(collection)
    search(collection)
```

This script does these things:
 1. inserts 5M entities
 2. searches without index
 3. builds index
 4. search with index, 

After running, we can see the search performance is different:
```
Number of entities in collection: 5000000
Load collection
Searching
Search time cost: 117.05923080444336 ms
Search result:
-["['(distance: 11.000551223754883, id: 6286)', '(distance: 11.03802490234375, id: 7588)', '(distance: 11.210186004638672, id: 7340)', '(distance: 11.352787971496582, id: 5306)', '(distance: 11.418453216552734, id: 4792)', '(distance: 11.436872482299805, id: 5794)', '(distance: 11.443241119384766, id: 6026)', '(distance: 11.517026901245117, id: 5351)', '(distance: 11.538634300231934, id: 8087)', '(distance: 11.572372436523438, id: 8262)']"]
Start building index
Index building state: {'total_rows': 5000000, 'indexed_rows': 5000000}
Release collection
Load collection
Searching
Search time cost: 13.873577117919922 ms
Search result:
-["['(distance: 12.100900650024414, id: 7446)', '(distance: 12.941367149353027, id: 9964)', '(distance: 12.987213134765625, id: 2915)', '(distance: 13.003852844238281, id: 9462)', '(distance: 13.149279594421387, id: 3749)', '(distance: 13.157496452331543, id: 9107)', '(distance: 13.162293434143066, id: 6125)', '(distance: 13.254480361938477, id: 1419)', '(distance: 13.526729583740234, id: 7920)', '(distance: 13.552371978759766, id: 4841)']"]

```

**Reply by martinenkoEduard** [reply]
Good example!)
